pubsub/kafkapubsub: often seems to hang on Travis

Or maybe just take too long? I've seen a lot of builds end on the Kafka tests with the "no response received for 10 minutes" message from Travis.

I've made some improvements so that the tests fail faster and with more information, in #2484 and #2482.

It looks like sometimes we never get the first message:

```
*** -> 0 got msg, acking: 990
*** calling Receive #999 for sub 0
got msg 998/1000 (after 1.145781405s)
*** -> 0 got msg, acking: 997
*** calling Receive #1000 for sub 0
got msg 999/1000 (after 1.146886088s)
*** 0 sending err to done
*** 0 returning
got premature error context deadline exceeded (after 30.000476343s)
*** didn't get msg 0
--- FAIL: TestMultiplePartionsWithRebalancing (33.08s)
    kafka_test.go:391: context deadline exceeded
```

--- a/pubsub/kafkapubsub/kafka.go
+++ b/pubsub/kafkapubsub/kafka.go
@@ -315,10 +315,11 @@ type subscription struct {
 	closeErr      error         // fatal error detected by the background consumer
 	consumerGroup sarama.ConsumerGroup
 
-	mu      sync.Mutex
-	unacked []*ackInfo
-	sess    sarama.ConsumerGroupSession // current session, if any, used for marking offset updates
-	claims  []sarama.ConsumerGroupClaim // claims in the current session
+	mu             sync.Mutex
+	unacked        []*ackInfo
+	sess           sarama.ConsumerGroupSession // current session, if any, used for marking offset updates
+	expectedClaims int                         // # of expected claims for the current session, they should be added via ConsumeClaim
+	claims         []sarama.ConsumerGroupClaim // claims in the current session
 }
 
 // ackInfo stores info about a message and whether it has been acked.
@@ -410,15 +411,14 @@ func openSubscription(brokers []string, config *sarama.Config, group string, top
 // Setup implements sarama.ConsumerGroupHandler.Setup. It is called whenever
 // a new session with the broker is starting.
 func (s *subscription) Setup(sess sarama.ConsumerGroupSession) error {
-	// The first time, close joinCh to (possibly) wake up OpenSubscription.
-	if s.joinCh != nil {
-		close(s.joinCh)
-		s.joinCh = nil
-	}
 	// Record the current session.
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.sess = sess
+	s.expectedClaims = 0
+	for _, claims := range sess.Claims() {
+		s.expectedClaims += len(claims)
+	}
 	return nil
 }
 
@@ -428,6 +428,7 @@ func (s *subscription) Cleanup(sarama.ConsumerGroupSession) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.sess = nil
+	s.expectedClaims = 0
 	s.claims = nil
 	return nil
 }
@@ -437,6 +438,11 @@ func (s *subscription) Cleanup(sarama.ConsumerGroupSession) error {
 func (s *subscription) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
 	s.mu.Lock()
 	s.claims = append(s.claims, claim)
+	// Once all of the expected claims have registered, close joinCh to (possibly) wake up OpenSubscription.
+	if s.joinCh != nil && len(s.claims) == s.expectedClaims {
+		close(s.joinCh)
+		s.joinCh = nil
+	}
 	s.mu.Unlock()
 	<-sess.Context().Done()
 	return nil

--- a/pubsub/kafkapubsub/kafka_test.go
+++ b/pubsub/kafkapubsub/kafka_test.go
@@ -339,9 +339,9 @@ func TestMultiplePartionsWithRebalancing(t *testing.T) {
 	send()
 
 	// Receive the messages via the subscription.
-	got := make(chan struct{})
+	got := make(chan int)
 	done := make(chan error)
-	read := func(ctx context.Context, sub *pubsub.Subscription) {
+	read := func(ctx context.Context, subNum int, sub *pubsub.Subscription) {
 		for {
 			m, err := sub.Receive(ctx)
 			if err != nil {
@@ -354,13 +354,13 @@ func TestMultiplePartionsWithRebalancing(t *testing.T) {
 				return
 			}
 			m.Ack()
-			got <- struct{}{}
+			got <- subNum
 		}
 	}
 	// The test will hang here if the messages aren't available, so use a shorter
 	// timeout.
 	ctx2, cancel := context.WithTimeout(ctx, 30*time.Second)
-	go read(ctx2, sub)
+	go read(ctx2, 0, sub)
 	for i := 0; i < nMessages; i++ {
 		select {
 		case <-got:
@@ -395,11 +395,13 @@ func TestMultiplePartionsWithRebalancing(t *testing.T) {
 
 	// The test will hang here if the message isn't available, so use a shorter timeout.
 	ctx3, cancel := context.WithTimeout(ctx, 30*time.Second)
-	go read(ctx3, sub)
-	go read(ctx3, sub2)
+	go read(ctx3, 0, sub)
+	go read(ctx3, 1, sub2)
+	counts := []int{0, 0}
 	for i := 0; i < nMessages; i++ {
 		select {
-		case <-got:
+		case sub := <-got:
+			counts[sub]++
 		case err := <-done:
 			// Premature error.
 			if err != nil {
@@ -412,6 +414,9 @@ func TestMultiplePartionsWithRebalancing(t *testing.T) {
 		if err := <-done; err != nil {
 			t.Fatal(err)
 		}
+	}
+	if counts[0] == 0 || counts[1] == 0 {
+		t.Errorf("one of the partitioned subscriptions didn't get any messages: %v", counts)
 	}
 }
 

